### PR TITLE
Distributed training example review

### PR DESCRIPTION
**Pull request by justinvyu**

### CODE REVIEW

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 1]
```diff
@@ -0,0 +1,962 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Enable Ray Train v2 (it's too good to wait for public release!)\n",
+    "ray.init(\n",
+    "    runtime_env={\n",
+    "        \"env_vars\": {\"RAY_TRAIN_V2_ENABLED\": \"1\"}, \n",
+    "        \"working_dir\": \"/home/ray/default\",  # to import doggos (default working_dir=\".\")\n",
+    "    },\n",
+    ")"
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:59]: 🤩 

No need to set the `RAY_TRAIN_V2_ENABLED=1` env var in the ray.init runtime env (Ray Train will do this automatically). It's just important to set this on the *driver* at the very beginning, before any `ray.train` imports. See "How to enable Train V2" in this guide: https://github.com/ray-project/ray/issues/49454

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 2]
```diff
@@ -0,0 +1,962 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "# This will be removed once Ray Train v2 is part of latest Ray version\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:69]:
```suggestion
    "# This will be removed once Ray Train v2 is enabled by default.\n",
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 3]
```diff
@@ -0,0 +1,962 @@
...
+   "metadata": {},
+   "source": [
+    "### Preprocess"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We need to convert our classes to labels (unique integers) so that we can train a classifier that can correctly predict the class given an input image. But before we do this, we'll quickly apply the same data ingestion and preprocessing as the previous notebook."
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:95]: Q: Are the notebooks required to be gone through in order? May want them to be more independent

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 4]
```diff
@@ -0,0 +1,962 @@
...
+    "    def convert_to_label(self, row, class_to_label):\n",
+    "        if \"class\" in row:\n",
+    "            row[\"label\"] = class_to_label[row[\"class\"]]\n",
+    "        return row\n",
+    "    \n",
+    "    def transform(self, ds, concurrency=4, batch_size=64, num_gpus=1):\n",
+    "        ds = ds.map(\n",
+    "            self.convert_to_label, \n",
+    "            fn_kwargs={\"class_to_label\": self.class_to_label},\n",
+    "        )\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:169]: I'm not a fan of using an instance method as the map batches UDF, since it implicitly captures the entire object `self` and can lead to some unexpected slowness/OOM if you accidentally save a huge object in this object, especially if your map doesn't even need that attribute.

Instead, explicitly pass the necessary state to the UDF (which is pretty close to what we do already):

```python
def convert_to_label(row, class_to_label):
    ...

ds = ds.map(convert_to_label, fn_args=(self.class_to_label,))
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 5]
```diff
@@ -0,0 +1,962 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class Preprocessor:\n",
+    "    \"\"\"Preprocessor class.\"\"\"\n",
+    "    def __init__(self, class_to_label=None):\n",
+    "        self.class_to_label = class_to_label or {}  # mutable defaults\n",
+    "        self.label_to_class = {v: k for k, v in self.class_to_label.items()}\n",
+    "        \n",
+    "    def fit(self, ds, column):\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:154]:
```suggestion
    "    def fit(self, ds, label_column):\n",
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 6]
```diff
@@ -0,0 +1,962 @@
...
+    "train_ds.take(1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert\"> <b> Data Processing</b> \n",
+    "\n",
+    "Be sure to checkout this extensive guide on [data loading and preprocessing](https://docs.ray.io/en/latest/train/user-guides/data-loading-preprocessing.html) for the last-mile preprocessing we'll need to do prior to training our models. However, Ray Data does support performant joins, filters, aggregations, etc. for the more structure data processing your workloads may need."
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:206]:
```suggestion
    "Be sure to checkout this extensive guide on [data loading and preprocessing](https://docs.ray.io/en/latest/train/user-guides/data-loading-preprocessing.html) for the last-mile preprocessing we'll need to do prior to training our models. Ray Data supports performant joins, filters, aggregations, etc. for the structured data processing your workloads may need."
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 7]
```diff
@@ -0,0 +1,962 @@
...
+    "\n",
+    "        # Define layers\n",
+    "        self.fc1 = nn.Linear(embedding_dim, hidden_dim)\n",
+    "        self.batch_norm = nn.BatchNorm1d(hidden_dim)\n",
+    "        self.relu = nn.ReLU()\n",
+    "        self.dropout = nn.Dropout(dropout_p)\n",
+    "        self.fc2 = nn.Linear(hidden_dim, num_classes)\n",
+    "\n",
+    "    def forward(self, batch):\n",
+    "        z = self.fc1(batch[\"embedding\"])\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:299]: It's a bit unclear that this batch is a dict of tensors at this point, since the dict format is an implicit Ray Data format, and the collate fn definition comes after. Maybe just add some text to say that the batch collation / formatting logic will be explained further down.

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 8]
```diff
@@ -0,0 +1,962 @@
...
+    "- And it also integrates with Ray's cluster launcher for cloud (AWS, GCP, K8s) and on-prem clusters. \n",
+    "- Solutions like PyTorch DDP require manually setting up your own process group, ranks, networking, etc.\n",
+    "\n",
+    "**🩹 2. Built-in fault tolerance**\n",
+    "- Ray Train supports automatic retry of failed workers.\n",
+    "- and can continue training from the last checkpoint in case of failure.\n",
+    "\n",
+    "\n",
+    "**✂️ 3. Flexible training strategies** (not just DDP)\n",
+    "- Ray Train supports Data Parallel, Model Parallel, Parameter Server, and even custom strategies.\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:682]:
```suggestion
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 9]
```diff
@@ -0,0 +1,962 @@
...
+    "- You can also use Torch DDP, FSPD, DeepSpeed, etc. under the hood if you want.\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:683]:
```suggestion
    "- Ray Train supports Torch DDP, FSDP, DeepSpeed, etc. with few to no code changes.\n",
```

Link to examples/quickstarts in docs.

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 10]
```diff
@@ -0,0 +1,962 @@
...
+    "- [Ray Compiled graphs](https://docs.ray.io/en/latest/ray-core/compiled-graph/ray-compiled-graph.html) allow us to even define different parallelism for jointly optimizing multipe models (Megatron, Deepspeed, etc. only allow for one global setting).\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:684]:
```suggestion
```

Probably too advanced for an intro guide, and this is largely unproven in terms of benefits.

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 11]
```diff
@@ -0,0 +1,962 @@
...
+    "\n",
+    "**🎛️ Multi-node orchestration made easy**\n",
+    "\n",
+    "- Ray Train automatically handles multi-node, multi-GPU setup with no manual SSH setup or hostfile configs. \n",
+    "- And it also integrates with Ray's cluster launcher for cloud (AWS, GCP, K8s) and on-prem clusters. \n",
+    "- Solutions like PyTorch DDP require manually setting up your own process group, ranks, networking, etc.\n",
+    "\n",
+    "**🩹 2. Built-in fault tolerance**\n",
+    "- Ray Train supports automatic retry of failed workers.\n",
+    "- and can continue training from the last checkpoint in case of failure.\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:678]: Also, "- leverages Ray autoscaler to request new replacement nodes in the event of node failures"

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 12]
```diff
@@ -0,0 +1,962 @@
...
+    "\n",
+    "\n",
+    "**✂️ 3. Flexible training strategies** (not just DDP)\n",
+    "- Ray Train supports Data Parallel, Model Parallel, Parameter Server, and even custom strategies.\n",
+    "- You can also use Torch DDP, FSPD, DeepSpeed, etc. under the hood if you want.\n",
+    "- [Ray Compiled graphs](https://docs.ray.io/en/latest/ray-core/compiled-graph/ray-compiled-graph.html) allow us to even define different parallelism for jointly optimizing multipe models (Megatron, Deepspeed, etc. only allow for one global setting).\n",
+    "\n",
+    "**🔥 Better support for heterogeneous clusters**\n",
+    "- Ray Train lets you define per-worker resource requirements (e.g., 2 CPUs and 1 GPU per worker).\n",
+    "- and can run on heterogeneous machines and scale flexibly (e.g., CPU for preprocessing and GPU for training)\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:688]:
```suggestion
    "- and can run on heterogeneous machines and scale flexibly (e.g., CPU nodes for dataset preprocessing and GPU for training)\n",
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 13]
```diff
@@ -0,0 +1,962 @@
...
+    "\n",
+    "**🌍 Integrations**\n",
+    "\n",
+    "<img src=\"https://raw.githubusercontent.com/anyscale/foundational-ray-app/refs/heads/main/images/train_integrations.png\" width=500>\n",
+    "\n",
+    "[RayTurbo Train](https://docs.anyscale.com/rayturbo/rayturbo-train) offers even more improvement to the price-performance ratio, performance monitoring and more:\n",
+    "- **elastic training** to ccale to a dynamic number of workers, continue training on fewer resources (even on spot instances).\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:695]:
```suggestion
    "- **elastic training**  which allows training to be started or resumed on a partial set of workers while waiting for other nodes to become available.\n",
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 14]
```diff
@@ -0,0 +1,962 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can view our experiment metrics and model artifacts in our model registry. We're using OSS mlflow so we can run the server by pointing to our model registry location:\n",
+    "\n",
+    "```bash\n",
+    "mlflow server -h 0.0.0.0 -p 8080 --backend-store-uri /mnt/user_storage/mlflow/doggos\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:721]: Mention that this is a network attached storage that can be accessed by all worker nodes.

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 15]
```diff
@@ -0,0 +1,962 @@
...
+   "metadata": {},
+   "source": [
+    "### Evaluation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll conclude by evaluating our trained model on our test dataset. Evaluation is essentially just the same as our batch inference workload -- where we'll apply the model on batches of data and then calculate metrics using the predictions vs.true labels. Ray data is hyper optimized for throughput so preserving order is not a priority. But for evaluation, this is crucial! So we'll achieve this by preserving the entire row and adding the predicted label as another column to each row."
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:794]: I don't think we need this paragraph -- there's no concern about labels not matching up with their image. Each row contains the right label corresponding to the image.

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 16]
```diff
@@ -0,0 +1,962 @@
...
+    "\n",
+    "**🔥 Better support for heterogeneous clusters**\n",
+    "- Ray Train lets you define per-worker resource requirements (e.g., 2 CPUs and 1 GPU per worker).\n",
+    "- and can run on heterogeneous machines and scale flexibly (e.g., CPU for preprocessing and GPU for training)\n",
+    "\n",
+    "**🌍 Integrations**\n",
+    "\n",
+    "<img src=\"https://raw.githubusercontent.com/anyscale/foundational-ray-app/refs/heads/main/images/train_integrations.png\" width=500>\n",
+    "\n",
+    "[RayTurbo Train](https://docs.anyscale.com/rayturbo/rayturbo-train) offers even more improvement to the price-performance ratio, performance monitoring and more:\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:694]:
```suggestion
    "[RayTurbo Train](https://docs.anyscale.com/rayturbo/rayturbo-train) offers even more improvement to the price-performance ratio, observability, and more:\n",
```

Comment (justinvyu) [reply notebooks/02-Distributed-Training_comments.ipynb:694]: "RayTurbo Train on Anyscale"

The dashboard mentioned later isn't part of "RayTurbo" but is on the Anyscale platform. Does this distinction matter?

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 17]
```diff
@@ -0,0 +1,962 @@
...
+    "- Ray Train automatically handles multi-node, multi-GPU setup with no manual SSH setup or hostfile configs. \n",
+    "- And it also integrates with Ray's cluster launcher for cloud (AWS, GCP, K8s) and on-prem clusters. \n",
+    "- Solutions like PyTorch DDP require manually setting up your own process group, ranks, networking, etc.\n",
+    "\n",
+    "**🩹 2. Built-in fault tolerance**\n",
+    "- Ray Train supports automatic retry of failed workers.\n",
+    "- and can continue training from the last checkpoint in case of failure.\n",
+    "\n",
+    "\n",
+    "**✂️ 3. Flexible training strategies** (not just DDP)\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:681]:
```suggestion
    "**✂️ 3. Flexible training strategies**\n",
```

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 18]
```diff
@@ -0,0 +1,962 @@
...
+    "    val_ds = ray.train.get_dataset_shard(\"val\")\n",
+    "\n",
+    "    # Model\n",
+    "    model = ClassificationModel(\n",
+    "        embedding_dim=embedding_dim, \n",
+    "        hidden_dim=hidden_dim, \n",
+    "        dropout_p=dropout_p, \n",
+    "        num_classes=num_classes,\n",
+    "    )\n",
+    "    model = ray.train.torch.prepare_model(model)\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:590]: Let's actually just wrap the model directly with `DistributedDataParallel` instead of using this ray train utility. We are planning to de-emphasize this API.

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 19]
```diff
@@ -0,0 +1,962 @@
...
+   "outputs": [],
+   "source": [
+    "def train_epoch(ds, batch_size, model, num_classes, loss_fn, optimizer):\n",
+    "    model.train()\n",
+    "    loss = 0.0\n",
+    "    ds_generator = ds.iter_torch_batches(batch_size=batch_size, collate_fn=collate_fn)\n",
+    "    for i, batch in enumerate(ds_generator):\n",
+    "        optimizer.zero_grad()  # reset gradients\n",
+    "        z = model(batch)  # forward pass\n",
+    "        targets = F.one_hot(batch[\"label\"], num_classes=num_classes).float()\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:522]: nit: I think cross entropy can take in just the batch["label"] index directly. No need for the explicit onehot conversion

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 20]
```diff
@@ -0,0 +1,962 @@
...
+    "        embedding_dim=embedding_dim, \n",
+    "        hidden_dim=hidden_dim, \n",
+    "        dropout_p=dropout_p, \n",
+    "        num_classes=num_classes,\n",
+    "    )\n",
+    "    model = ray.train.torch.prepare_model(model)\n",
+    "\n",
+    "    # Training components\n",
+    "    loss_fn = torch.nn.CrossEntropyLoss()\n",
+    "    optimizer = torch.optim.Adam(model.parameters(), lr=lr)\n",
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:594]: nit: AdamW

File: notebooks/02-Distributed-Training_comments.ipynb [hunk 21]
```diff
@@ -0,0 +1,962 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Scaling config\n",
+    "num_workers = 4\n",
+    "scaling_config = ray.train.ScalingConfig(\n",
+    "    num_workers=num_workers,\n",
+    "    use_gpu=True,\n",
+    "    resources_per_worker={\"CPU\": 8, \"GPU\": 1})"
```
Comment (justinvyu) [inline notebooks/02-Distributed-Training_comments.ipynb:494]:
```suggestion
```

Setting resources_per_worker is an advanced configuration that's not needed here (it just steals logical CPUs from Ray Data).

**Review by justinvyu — COMMENTED** [review]
Nice!

File: notebooks/02-Distributed-Training_comments.ipynb (hunk 4 above)
Comment (GokuMohandas) [reply notebooks/02-Distributed-Training_comments.ipynb:169]: fantastic point Justin!
